### PR TITLE
dev-python/cryptography: the snapshot of the new version with rust

### PR DESCRIPTION
**Pull request by vaartis**
Separated from #20795. The new version of cryptography now requires rust. Putting this up here in case we need it in the future.

### CODE REVIEW

**Comment by gyakovlev** [comment]
there's practically no rust extention in 3.4.7
https://github.com/pyca/cryptography/blob/3.4.7/src/rust/src/lib.rs

**Comment by vaartis** [comment]
> there's practically no rust extention in 3.4.7
> https://github.com/pyca/cryptography/blob/3.4.7/src/rust/src/lib.rs

This isn't 3.4.7, but a snapshot of the latest version from github. 3.4.7, which we already have, has the ability to disable rust, which we do. After that release, they removed the ability to disable rust and started converting C code to rust. Initially this was made when I worked on porting to python 3.10, but it turned out backporting a patch was enough for that. If we ever chose to support the newer versions which, if the developers of the library do not decide otherwise or someone doesn't fork the library, will require rust, this snapshot ebuild could be the starting point for this new version.

**Comment by gyakovlev** [comment]
> > there's practically no rust extention in 3.4.7
> > [https://github.com/pyca/cryptography/blob/3.4.7/src/rust/src/lib.rs](https://github.com/pyca/cryptography/blob/3.4.7/src/rust/src/lib.rs?rgh-link-date=2021-05-15T07%3A48%3A24Z)
> 
> This isn't 3.4.7, but a snapshot of the latest version from github. 3.4.7, which we already have, has the ability to disable rust, which we do. After that release, they removed the ability to disable rust and started converting C code to rust. Initially this was made when I worked on porting to python 3.10, but it turned out backporting a patch was enough for that. If we ever chose to support the newer versions which, if the developers of the library do not decide otherwise or someone doesn't fork the library, will require rust, this snapshot ebuild could be the starting point for this new version.

Sorry I don't see snapshot, just a revbump with patch.
did you forget to push something or still working on it? I that case I might be early to review =)

**Comment by vaartis** [comment]
>  did you forget to push something or still working on it? I that case I might be early to review =) 
You are correct.. seems like i pushed the wrong commit. I'm going to push the correct one soon. I was suggested to make a draft PR for this since it's not known whether we will need this, but it may be useful in the future.

**Comment by sbraz** [comment]
There's a typo in the PR title: "cryptorgaphy".

File: dev-python/cryptography/cryptography-3.4.7_p12052021.ebuild
Comment (gyakovlev) [inline dev-python/cryptography/cryptography-3.4.7_p12052021.ebuild:89]: quote it.

File: dev-python/cryptography/cryptography-3.4.7_p12052021.ebuild
Comment (gyakovlev) [inline dev-python/cryptography/cryptography-3.4.7_p12052021.ebuild:113]: quote

File: dev-python/setuptools_rust/metadata.xml
Comment (gyakovlev) [inline dev-python/setuptools_rust/metadata.xml:12]: normalize xml with tabs
`XMLLINT_INDENT=$'\t' xmllint --format metadata.xml  > newname.xml`

File: dev-python/setuptools_rust/metadata.xml
Comment (gyakovlev) [inline dev-python/setuptools_rust/metadata.xml:7]: add `rust@gentoo.org` as second maintainer.

File: dev-python/cryptography/cryptography-3.4.7_p12052021.ebuild
Comment (gyakovlev) [inline dev-python/cryptography/cryptography-3.4.7_p12052021.ebuild:79]: sort

**Review by gyakovlev — CHANGES_REQUESTED** [review]
looks good, left some comments.

**Comment by gyakovlev** [comment]
also I think it's ok to add `<stabilize-allarches/>` to `setuptools_rust` as it's python only code.

**Comment by vaartis** [comment]
I think I got it all. Also rebased on master, where the older cryptography version was removed.

File: dev-python/setuptools_rust/setuptools_rust-0.12.1.ebuild
Comment (gyakovlev) [inline dev-python/setuptools_rust/setuptools_rust-0.12.1.ebuild:21]: ditto with keywords here. only add `~amd64`, the bot check will become green.

Comment (vaartis) [reply dev-python/setuptools_rust/setuptools_rust-0.12.1.ebuild]: Right, of course. Done.

**Comment by vaartis** [comment]
Unfortunately there's still a semantic_version bump to 3.10 in another PR that is needed for this one, so the warnings won't go away until it's merged.

**Comment by gyakovlev** [comment]
ok let's wait then until rest is processed, thanks for working on it btw. setuptoos_rust was my on my to-do list, now it's not ;-)

**Comment by vaartis** [comment]
Rebased on master since semantic_version got merged. This should make the QA bot happy.

**Comment by gyakovlev** [comment]
found time to test.

you can add `~ppc64` keywords, tested by me, tests pass.
`2722 passed, 3252 skipped, 86631 warnings in 241.48s (0:04:01) `

`RUSTFLAGS` seem to work fine,  good.

qa warning 1 for `setuptools_rust`

```
 * 
 * QA Notice: this package seems to contain tests but they are not enabled.
 * Please either run tests (via distutils_enable_tests or declaring
 * python_test yourself), or add RESTRICT="test" along with an explanatory
 * comment if tests cannot be run.
 * 
 ```
 

qa warning 2 for `cryptography`

```
* QA Notice: Files built without respecting CFLAGS have been detected
 *  Please include the following list of files in your report:
 * /usr/lib/python3.9/site-packages/cryptography/hazmat/bindings/_rust.abi3.so


 * QA Notice: Files built without respecting LDFLAGS have been detected
 *  Please include the following list of files in your report:
 * /usr/lib/python3.9/site-packages/cryptography/hazmat/bindings/_rust.abi3.so
 ```
probably 
`QA_FLAGS_IGNORED="usr/lib.*/py.*/site-packages/cryptography/hazmat/bindings/_rust.abi3.so"`
should shut it up, it's acceptable for rust.
 

also `src_unpack` spams with 
```
>>> Unpacking winapi-0.3.9.crate to /var/tmp/portage/dev-python/cryptography-3.4.7_p12052021/work                                                                                                                                                                                     
unpack winapi-0.3.9.crate: file format not recognized. Ignoring.
....
```

you should not call `default` in `src_unpack` to avoid it, `cargo_src_unpack` alone is enough and will handle tarballs just fine.
unknown formats in `SRC_URI` passed to unpack  may become errors in future `EAPI`.

otherwise it's good =)

**Comment by vaartis** [comment]
OK, so I _sort of_ got `setuptools_rust` "tests" to work: they use the examples dir with several rust things for testing. Unfortunately two of them I was not able to get to work. html_py_ever works if I run the same commands as portage does by hand, but when running with emerge it can't import the module it's looking for, and rust_with_cffi uses _a different version of pyo3_ for some reason and i think cargo.eclass can't handle that (maybe because it's not even specified in cargo.toml, but instead is marked as a transitive dependency and `cargo ebuild` doesn't even show it).

**Comment by thesamesam** [comment]
(Try using distutils_install_for_testing with its various options?)

**Comment by gyakovlev** [comment]
I don't think examples are real tests.
just add 
`RESTRICT=test #no tests`
is sucks ofc, but we tried.

**Comment by gyakovlev** [comment]
but I did not look at examples, maybe worth keeping. I'll give it another go.

**Comment by vaartis** [comment]
> but I did not look at examples, maybe worth keeping. I'll give it another go.

The examples have test in them, and are considered "tests" by upstream.

**Comment by vaartis** [comment]
@gyakovlev, i was able to get html-py-ever to work, plus worked out the annoying thing of it trying to install the built examples into the system. unforunately the src_test phase is a bit complex, becase the tests are super inconsistent, but it works. still no luck with `rust_with_cffi` unfortunately, i don't think we can do anything about that. but `html-py-ever` seems to be the biggest real test example.

**Comment by gyakovlev** [comment]
thanks.
I might have an idea how to fix different cargo crates.
we could define a var, which lists deps required by `rust_with_cffi`
```
TEST_CRATES="
...
"
```

and extend `SRC_URI` with

```
...
    test?  ( $(cargo_crate_uris ${TEST_CRATES}) )
```

but again, haven't looked at it in detail, it's just an assumption based on fact the example needs different crate version.

**Comment by gyakovlev** [comment]
I got lost here, seems it's what you've done already.

**Comment by vaartis** [comment]
> I got lost here, seems it's what you've done already.

Yes, but if I list both versions there I think only one gets downloaded by cargo.eclass..

**Comment by gyakovlev** [comment]
I see where problem is.
version `0.12.1` of `setuptools_rust` requires git checkout of pyo3, hence it's not handled by eclass/cargo-ebuild.

current main branch switched versioned pyo3, so it should work in the future.

https://github.com/PyO3/setuptools-rust/commit/21fc0105dc7ba9b3bfdce1530a80190c01dff560#diff-a3f14f562689def4c1df269f3e37bafcf6036a58798d825b8a58438cb479600e

you can list multiple versions of pyo3 in `CRATES`, that will be fine btw.

also, cargo-ebuild was a bit out of maintenance. it may fail to catch all crates.
I recently taken over maintenance of it and will fix it soon.

what I use is right now is
```
CARGO_HOME=/tmp/crates cargo fetch
ls -1 /tmp/crates/registry/src/github.com-*/
```

it will give you a bit more reliable list of crates.

to really catch all deps of all examples - just repeat `fetch` step for all examples containing cargo.toml and `ls -1` shoud contain all crates that can be possibly used. 
as bonus point - you can copy downloaded crates from `/tmp/crates/registry/cache/github.com-*/` to your distdir.
and ofc clean-up /tmp/crates after that and before run too.

**Comment by gyakovlev** [comment]
```
python3.9 -m pytest -vv -ra -l
/usr/bin/python3.9: No module named pytest
```

looks like setuptools_rust needs pytest test dep.

**Comment by gyakovlev** [comment]
after adding pytest it passes.
```
tests/test_namespace_package.py::test_rust PASSED                                                                                   [ 50%]
tests/test_namespace_package.py::test_cffi PASSED                                                                                   [100%]
```
we should look at merging setuptools_rust tbh and leave crypto in Draft for now.

**Comment by vaartis** [comment]
Yes, it does. Sorry, forgot to add it.

**Comment by vaartis** [comment]
>current main branch switched versioned pyo3, so it should work in the future.
> you can list multiple versions of pyo3 in CRATES, that will be fine btw.

Ah, i see why it couldn't fetch it then. my bad.

**Comment by vaartis** [comment]
I guess we still can't use the crate from git easily, so that's for the next release.

**Comment by vaartis** [comment]
The html-py-ever tests actually require beautifulsoup and it hasn't been ported to 3.10 yet, it has some problems with compatibility unfortunately. Didn't notice initially because I have it installed in 3.10 locally.

**Comment by gyakovlev** [comment]
merged `setuptools_rust`, and committed just a bit of fixes in https://github.com/gentoo/gentoo/commit/6f304b0a6d911510c647e880df2c5efc3c13323f
don't want to nitpick it anymore. you've done a lot already.

**Comment by vaartis** [comment]
Thank you. Should I watch when the next release comes out and update this?

**Comment by gyakovlev** [comment]
> Thank you. Should I watch when the next release comes out and update the package then?

as you wish, actual help is always appreciated =)

**Comment by gyakovlev** [comment]
something to put into commit message:

`Bug: https://bugs.gentoo.org/769482`

**Comment by thesamesam** [comment]
Thanks a lot @vaartis! Merged and then updated to 36.0.0.

**Pull request closed without merging**
